### DISCUSSION TITLE
Named window functions

### DISCUSSION THREAD

**Discussion by lukasjuhrich** [sqlalchemy/sqlalchemy#8350] [Ideas]
Currently, I did not find a way to produce a statement with a named `WINDOW w AS …` like in the following query:

```sql
select
  m.user_id,
  last_value(m.id) over w,
  last_value(upper(m.active_during)) over w
from membership m
window w as (partition by user_id order by upper(m.active_during) nulls last);
```

<details><summary>Table Schema</summary>

```sql
create table membership (
    id serial primary key,
    user_id integer not null,
    group_id integer not null,
    active_during tstzrange not null
);
```
</details>

Would it be possible to introduce syntax like
```python
select()\
.select_from(Membership)\
.add_window(
    "w",
    partition_by=Membership.user_id,
    order_by=nulls_last(mem_ends_at),
)\
.add_columns(
    Membership.user_id,
    Membership.active_during.over(w),
    Membership.id.over(w),
)
```

As an alternative to the following redundant statement?
```python
select()\
.select_from(Membership)\
.add_columns(
    Membership.user_id,
    Membership.active_during.over(
        partition_by=Membership.user_id,
        order_by=nulls_last(mem_ends_at),
    ),
    Membership.id.over(
        partition_by=Membership.user_id,
        order_by=nulls_last(mem_ends_at),
    ),
)
```

**Comment by zzzeek** [comment]
hi - 

Im completely unfamiliar with this syntax or what database and version happens to have this.  if you could provide background on where this comes from that would be helpful (I'm sure it's PostgreSQL but please point it out for me).

**Reply by lukasjuhrich** [reply]
Good evening,

I did not manage to get the official SQL standard, but I think this might not actually be postgres specific:
- The [wiki article](https://en.wikipedia.org/wiki/Window_function_(SQL)#Syntax) just copies the PostgreSQL syntax
- However, it claims introduction in `SQL:2003` while linkink to [this mariadb documentation](https://mariadb.com/kb/en/window-functions-overview/), where someone mentions use of named window functions in the comments.
- **Postgresql** seems to have introduced the syntax since the initial window function implementation in **v8.4** ([`SELECT` v8.3](https://www.postgresql.org/docs/8.3/sql-select.html) / [`SELECT` v8.4](https://www.postgresql.org/docs/8.4/sql-select.html), [changelog](https://www.postgresql.org/docs/release/8.4.0/) just says „Add support for WINDOW functions“)

Let me see whether I can get my hands on `SQL:2003` somehow.

**Reply by zzzeek** [reply]
> Good evening,
> 
> I did not manage to get the official SQL standard, but I think this might not actually be postgres specific:
> 
>     * The [wiki article](https://en.wikipedia.org/wiki/Window_function_(SQL)#Syntax) just copies the PostgreSQL syntax

OK, we support that syntax, that's just ordinary window functions that are used in the columns clause of the SELECT.  The SQL you illustrated at the top shows the window function in the FROM clause with an alias name.  There's ways to make that happen also but I'd need to see a real working query.

if you are just looking for window functions the way the example at https://en.wikipedia.org/wiki/Window_function_(SQL)#Syntax illustrates, i.e.

```sql
 SELECT
  LAG(name, 1) 
    OVER(ORDER BY name) "prev",
  name, 
  LEAD(name, 1) 
    OVER(ORDER BY name) "next"
 FROM people
 ORDER BY name
```

the tutorial goes over how to render SQL of this style [here](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#using-window-functions).

**Reply by ohaibbq** [reply]
Hi @zzzeek, here's a query that utilizes the named window function syntax [from the Postgresql docs](https://www.postgresql.org/docs/current/tutorial-window.html).

> When a query involves multiple window functions, it is possible to write out each one with a separate OVER clause, but this is duplicative and error-prone if the same windowing behavior is wanted for several functions. Instead, each windowing behavior can be named in a WINDOW clause and then referenced in OVER. For example:

```sql
SELECT 
  sum(salary) OVER w,
  avg(salary) OVER w
FROM empsalary
WINDOW w AS (PARTITION BY depname ORDER BY salary DESC);
```

---

Other SQLAlchemy APIs are built in a way that you can extract duplicate clauses into a variable, then use the variable in multiple places of the query. This is not really possible for windows, you instead need to extract at one layer of abstraction higher. For example, you could extract the `**kwargs` passed to a `func.over()` call, or extract a function that produces an `over(clause, **kwargs)` expression.

Example of the current API:
```python
by_department = dict(partition_by=['depname'], order_by=['salary DESC'])
results = (session.query(
  func.sum(EmployeeSalary.salary).over(**by_department),
  func.avg(EmployeeSalary.salary).over(**by_department)
))
```

Potential APIs:
```python
by_department = window("by_department", partition_by=['depname'], order_by=['salary DESC'])
results = (session.query(
  func.sum(EmployeeSalary.salary).over(by_department),
  func.avg(EmployeeSalary.salary).over(by_department)
))

results = (session.query(
  func.sum(EmployeeSalary.salary).over('by_department'),
  func.avg(EmployeeSalary.salary).over('by_department')
).window(by_department))
```

It would be nice to have support for named functions, if for nothing else than improved readability when looking at the compiled SQL queries.

--

edit: just saw the thread in #8352 :)

**Reply by zzzeek** [reply]
the endmost proposed syntax is at https://github.com/sqlalchemy/sqlalchemy/issues/8352#issuecomment-1205980904 so go for that one

**Comment by CaselIT** [comment]
since this is a feature request, an issue is probably best. let me open one

**Comment by zzzeek** [comment]
I've proposed what this syntax would look like at #8352 however there's no technical rationale for it right now.

for the purposes of composability, SQLAlchemy offers composability by using Python variables to represent SQL constructs.   If you compare my example of the proposed syntax at https://github.com/sqlalchemy/sqlalchemy/issues/8352#issuecomment-1205722128 to a current example at https://github.com/sqlalchemy/sqlalchemy/issues/8352#issuecomment-1205724074, the Python code matches on a line-by-line basis.   The SQL renders differently, but that's fine if the only difference is composability.
